### PR TITLE
Auto driver

### PR DESCRIPTION
**Pull request by OliverW10**
Follows a linear path between waypoints with a trapazoidal velocity profile.
It will follow a trapazoidal velocity profile to stop before the next waypoint if it is holding a ball otherwise will not stop

PR stack:

- #24
- #25 :point_left: this PR

### CODE REVIEW

File: utilities/trajectory_generator.py [hunk 1]
```diff
@@ -36,13 +37,24 @@ def smooth_path(
...
         x += sample_pose.X()
         y += sample_pose.Y()
-        angle += sample_pose.rotation().radians()
-    return Pose2d(x / sample_count, y / sample_count, angle / sample_count)
+        angle_x += sample_pose.rotation().cos()
+        angle_y += sample_pose.rotation().sin()
+    return Pose2d(x / sample_count, y / sample_count, math.atan2(angle_y, angle_x))
+
+
+def ours_to_wpi(pose: Pose2d):
```
Comment (auscompgeek) [inline utilities/trajectory_generator.py:55]: This is a confusing name. It'd be better if it actually described what it actually does, e.g. "robot pose to field"

File: utilities/trajectory_generator.py [hunk 2]
```diff
@@ -36,13 +37,24 @@ def smooth_path(
...
+    """Converts a pose in our goal centered system to
+    the corner center that wpilib uses (e.g. for field2d display)"""
+    field_x = 324 * 0.0254  # half field x dimension
+    field_y = 162 * 0.0254  # half field y dimension
+    return Pose2d(pose.X() + field_x, pose.Y() + field_y, pose.rotation())
```
Comment (auscompgeek) [inline utilities/trajectory_generator.py:60]: Should we add a `Transform2d` instead?

File: utilities/trajectory_generator.py [hunk 3]
```diff
@@ -36,13 +37,24 @@ def smooth_path(
...
             sample, 0, sample_count, -look_around, look_around
         )
         sample_pose = lookup_linear(waypoints, sample_d)
         x += sample_pose.X()
         y += sample_pose.Y()
-        angle += sample_pose.rotation().radians()
-    return Pose2d(x / sample_count, y / sample_count, angle / sample_count)
+        angle_x += sample_pose.rotation().cos()
+        angle_y += sample_pose.rotation().sin()
+    return Pose2d(x / sample_count, y / sample_count, math.atan2(angle_y, angle_x))
```
Comment (auscompgeek) [inline utilities/trajectory_generator.py:52]: Should this call Rotation2d instead of math.atan2?

File: autonomous/autonomous.py [hunk 4]
```diff
@@ -1,10 +1,256 @@
...
+        # applies to the linear speed, not turning
+        self.linear_constraints = TrapezoidProfile.Constraints(1.5, 1)
+        # since chassis speeds should be limited in trajectory generation this is high to not hold it back
+        self.drive_rotation_constrants = trajectory.TrapezoidProfileRadians.Constraints(
+            3, 5
+        )
+        self.drive_controller = controller.HolonomicDriveController(
+            controller.PIDController(5, 0, 0),
+            controller.PIDController(5, 0, 0),
+            controller.ProfiledPIDControllerRadians(
```
Comment (auscompgeek) [inline autonomous/autonomous.py:40]: do you need to `setContinuous` on the theta controller?

File: utilities/trajectory_generator.py [hunk 5]
```diff
@@ -1,21 +1,41 @@
-from wpimath.geometry import Pose2d
+from wpimath.geometry import Pose2d, Rotation2d, Translation2d
 from typing import List
 
 from utilities.scalers import scale_value
 
 
+field_x = 324 * 0.0254  # half field x dimension
+field_y = 162 * 0.0254  # half field y dimension
+FIELD_TRANFORM = Translation2d(field_x, field_y)
```
Comment (auscompgeek) [inline utilities/trajectory_generator.py:9]: This is no longer a transform, so should not be called as such.

Should this be called `GOAL_POSITION` or similar?

File: autonomous/autonomous.py [hunk 6]
```diff
@@ -1,10 +1,248 @@
...
+    """Auto starting middle of right tarmac, picking up balls 3, 2 and both at 4"""
+
+    MODE_NAME = "Five Ball"
+    DEFAULT = True
+
+    def __init__(self):
+        self.waypoints = [
+            start_positions[0],
+            Pose2d(-0.711, -3.5, -math.pi / 2),  # 3
+            Pose2d(-2.789, -2.378, math.radians(-206)),  # 2
```
Comment (auscompgeek) [inline autonomous/autonomous.py:216]: Would calling Rotation2d.fromDegrees be clearer?

File: autonomous/autonomous.py [hunk 7]
```diff
@@ -1,10 +1,248 @@
...
+    (-0.681, -3.826),  # right
+    (-7.156, -3.010),  # terminal
+    (-7.924, -1.559),  # cargo line
+]
+red_balls = [
+    (-2.249, 3.169),  # left
+    (-3.771, -0.935),  # middle
+    (-0.850, -3792),  # right
+]
+start_positions = [
```
Comment (auscompgeek) [inline autonomous/autonomous.py:187]: Any particular reason for this list? I don't see it being used as a list anywhere from a quick glance.

**Comment by OliverW10** [comment]
this now just follows a path of waypoints, no stopping, we will build on top of this adding more states later

**Review by LucienMorey — APPROVED** [review]

**Pull request merged**